scipy-1.0/paper.tex

```diff
@@ -258,6 +262,20 @@ \subsubsection*{LowLevelCallable}
 
 \subsection*{Cython bindings for BLAS, LAPACK and special}
 
+When writing algorithms or wrappers in Cython, it is often convenient to call directly into the libraries that SciPy wraps rather than routing the call back through the Python interpreter.
+SciPy currently includes Cython wrappers for many BLAS and LAPACK routines and for many of the special functions provided in the \texttt{scipy.{\allowbreak}special} submodule.
+These wrappers are available in the modules \texttt{scipy.{\allowbreak}linalg.{\allowbreak}cython\_blas}, \texttt{scipy.{\allowbreak}linalg.{\allowbreak}cython\_lapack}, and \texttt{scipy.{\allowbreak}special.{\allowbreak}cython\_special} respectively.
+The low-level interfaces for Cython can also be used by third party libraries to get direct access to the functions in these libraries while avoiding the overhead of Python function calls.
+Cython's cimport machinery allows Cython based modules to share C-level objects with other Cython modules without requiring that the third-party extensions link against any additional libraries.
+Function pointers to wrappers for library routines can also be shared in this way, so that Cython modules using the interface provided by SciPy no longer have to worry about finding and linking against the original library.
+This lets other extensions avoid the complexity of finding and using the correct libraries on the various possible Fortran compiler ABI and OS combinations.
+The low-level Cython wrappers provide an OS and Fortran compiler independent interface for the libraries they wrap.
+
+To help with both correctness and ease of maintenance, most of these low-level Cython wrappers are generated automatically.
+The wrappers for BLAS and LAPACK are primarily generated using type information that is parsed from the BLAS and LAPACK source files using f2py, though a small number of routines use hand-written type signatures instead.
+The input and output types of each routine are saved in a data file that is read at build time and used to generate the corresponding Cython wrapper files.
+The wrappers in \texttt{scipy.{\allowbreak}special.{\allowbreak}cython\_special} are also generated from a data file containing type information for the wrapped routines.
```

Comment:
Was the data file for `cython_special` not generated by `f2py`? Was it handwritten? If one of these, it's possible that the information can be slipped into one of the previous sentences rather than being a separate sentence (which could make it more specific/clear and shorter at the same time).

Reply:
Good question. I'm less familiar with the internal layout of `scipy.special` so I was being cautious. As I understand it, the functions there have two interfaces. One is the ufunc interface that can be used from within Python and the other is the low-level Cython interface that can only be used from within Cython. Both sets of wrappers are generated from a single data file (https://github.com/scipy/scipy/blob/master/scipy/special/functions.json) that describes the interface for each function being wrapped. I believe the information there is written by hand though. Perhaps @person142 could clarify more though.

Reply:
@person142 If you can answer this, I think it would help to merge this sentence with the one about BLAS and LAPACK.